service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -665,4 +666,37 @@ protected final Message newMessage(
         addProperties(msg, registrationInfo);
         return msg;
     }
+
+    /**
+     * Approve a message defined by a Content-Type and a payload for protocol adapters.
+     * @param contentType The Content-Type of the payload, as received by the protocol adapter.
+     * @param payload The payload received by the protocol adapter.
+     * @return A String containing an error message (for further processing), or an empty Optional if the message was
+     *         approved successfully.
+     */
+    protected Optional<String> approveMessage(final Optional<String> contentType, final Buffer payload) {
```

Comment:
FMPOV introducing this method was counter-productive. Code that was easy to read and understand before (the checks in the request handler methods) is now very hard to understand, not the least because the name of this method suggests some sort of *terminal* handling of the request.

FMPOV we should keep the original code for the sake of readability even if it means that there is some duplication.

Reply:
Resulting in something like 

```
        if (contentType == EventConstants.CONTENT_TYPE_EMPTY_NOTIFICATION && payload != null && payload.length() > 0) {
            HttpUtils.badRequest(ctx, String.format("Content-Type %s does not allow non empty payload", contentType));
        } else if (contentType != EventConstants.CONTENT_TYPE_EMPTY_NOTIFICATION && (payload == null || payload.length() == 0)) {
            HttpUtils.badRequest(ctx, String.format("Content-Type %s does not allow empty payload", contentType));
        } else {
```

I do not find that too readable either, to be honest, but I can live with that.
After rebase there will be several more else, maybe afterwards a simple extract refactoring makes sense.
Will adopt that then...

Reply:
Looking into it a little further I find that keeping the check in a separate method is much better testable. Now there are some easy to understand test methods for this that I would like to keep somehow.

So my question is:
do you dislike mainly the name of the method? Then it may help to just rename it to `checkPayloadByType`.
or do you dislike the usage of `Optional` for the return value to contain the error message? Then it could help to return boolean instead and let the adapter print a little more generic message for that.

Or do you think of a third option?

Reply:
I think it's a combination of both the name and the usage of the optional as the return value. When returning a boolean, the method should be named starting with an `is...` or `has...` like when asking a question, e.g. `hasPayloadOfIndicatedType`  ...

Reply:
This I can understand (although I liked the idea of using an Optional to return a `boolean` together with a possible error message, would be great if Java supported multiple return values like a lot of other languages now do...).

So I will use a `boolean isPayloadOfIndicatedType(...)` instead.